flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerServiceAsyncImpl.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.flink.util.function.BiConsumerWithException;
+
+import static org.apache.flink.runtime.asyncprocessing.KeyAccountingUnit.EMPTY_RECORD;
+
+/**
+ * An implementation of {@link InternalTimerService} that is used by {@link
+ * org.apache.flink.streaming.runtime.operators.asyncprocessing.AbstractAsyncStateStreamOperator}.
+ * The timer service will set {@link RecordContext} for the timers before invoking action to
+ * preserve the execution order between timer firing and records processing.
+ *
+ * @see <a
+ *     href=https://cwiki.apache.org/confluence/display/FLINK/FLIP-425%3A+Asynchronous+Execution+Model#FLIP425:AsynchronousExecutionModel-Timers>FLIP-425
+ *     timers section.</a>
+ * @param <K> Type of timer's key.
+ * @param <N> Type of the namespace to which timers are scoped.
+ */
+public class InternalTimerServiceAsyncImpl<K, N> extends InternalTimerServiceImpl<K, N> {
+
+    private AsyncExecutionController<K> asyncExecutionController;
+
+    InternalTimerServiceAsyncImpl(
+            TaskIOMetricGroup taskIOMetricGroup,
+            KeyGroupRange localKeyGroupRange,
+            KeyContext keyContext,
+            ProcessingTimeService processingTimeService,
+            KeyGroupedInternalPriorityQueue processingTimeTimersQueue,
+            KeyGroupedInternalPriorityQueue eventTimeTimersQueue,
+            StreamTaskCancellationContext cancellationContext,
+            AsyncExecutionController<K> asyncExecutionController) {
+        super(
+                taskIOMetricGroup,
+                localKeyGroupRange,
+                keyContext,
+                processingTimeService,
+                processingTimeTimersQueue,
+                eventTimeTimersQueue,
+                cancellationContext);
+        this.asyncExecutionController = asyncExecutionController;
+        this.processingTimeCallback = this::onProcessingTime;
+    }
+
+    private void onProcessingTime(long time) throws Exception {
+        // null out the timer in case the Triggerable calls registerProcessingTimeTimer()
+        // inside the callback.
+        nextTimer = null;
+
+        InternalTimer<K, N> timer;
+
+        while ((timer = processingTimeTimersQueue.peek()) != null
+                && timer.getTimestamp() <= time
+                && !cancellationContext.isCancelled()) {
+            RecordContext<K> recordCtx =
+                    asyncExecutionController.buildContext(EMPTY_RECORD, timer.getKey());
+            recordCtx.retain();
+            asyncExecutionController.setCurrentContext(recordCtx);
+            keyContext.setCurrentKey(timer.getKey());
+            processingTimeTimersQueue.poll();
+            final InternalTimer<K, N> timerToTrigger = timer;
+            asyncExecutionController.syncPointRequestWithCallback(
+                    () -> triggerTarget.onProcessingTime(timerToTrigger));
+            taskIOMetricGroup.getNumFiredTimers().inc();
+            recordCtx.release();
```

Comment:
Can we extract the common part of `onProcessingTime`, `advanceWatermark` and `foreachTimer` into one helper method? e.g. `maintainContextAndProcess(K key, ThrowingRunnable<Exception> runnable)`.